File: 2024/day01/node/solution.js
import { readFileSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __dirname = dirname(fileURLToPath(import.meta.url));
const input = readFileSync(join(__dirname, '..', 'input.txt'), 'utf-8').trim();

// Parse input
const lines = input.split('\n');

// Part 1
function part1() {
  const leftList = [];
  const rightList = [];

  for (const line of lines) {
    const [left, right] = line.split(/\s+/).map(Number);
    leftList.push(left);
    rightList.push(right);
  }

  // Sort both lists
  leftList.sort((a, b) => a - b);
  rightList.sort((a, b) => a - b);

  // Calculate total distance
  let totalDistance = 0;
  for (let i = 0; i < leftList.length; i++) {
    totalDistance += Math.abs(leftList[i] - rightList[i]);
  }

  return totalDistance;
}

// Part 2
function part2() {
  const leftList = [];
  const rightList = [];

  for (const line of lines) {
    const [left, right] = line.split(/\s+/).map(Number);
    leftList.push(left);
    rightList.push(right);
  }

  // Count occurrences in right list
  const rightCounts = new Map();
  for (const num of rightList) {
    rightCounts.set(num, (rightCounts.get(num) || 0) + 1);
  }

  // Calculate similarity score
  let similarityScore = 0;
  for (const num of leftList) {
    similarityScore += num * (rightCounts.get(num) || 0);
  }

  return similarityScore;
}

console.log('Part 1:', part1());
console.log('Part 2:', part2());

File: 2024/day01/java/Solution.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Solution {
    private static class InputLists {
        List<Integer> left;
        List<Integer> right;

        InputLists(List<Integer> left, List<Integer> right) {
            this.left = left;
            this.right = right;
        }
    }

    private static InputLists readInput() throws IOException {
        List<String> lines = Files.readAllLines(Paths.get("../input.txt"));
        List<Integer> leftList = new ArrayList<>();
        List<Integer> rightList = new ArrayList<>();

        for (String line : lines) {
            String[] parts = line.trim().split("\\s+");
            if (parts.length == 2) {
                leftList.add(Integer.parseInt(parts[0]));
                rightList.add(Integer.parseInt(parts[1]));
            }
        }

        return new InputLists(leftList, rightList);
    }

    private static int part1(List<Integer> leftList, List<Integer> rightList) {
        // Create copies to avoid modifying originals
        List<Integer> left = new ArrayList<>(leftList);
        List<Integer> right = new ArrayList<>(rightList);

        // Sort both lists
        Collections.sort(left);
        Collections.sort(right);

        // Calculate total distance
        int totalDistance = 0;
        for (int i = 0; i < left.size(); i++) {
            totalDistance += Math.abs(left.get(i) - right.get(i));
        }

        return totalDistance;
    }

    private static int part2(List<Integer> leftList, List<Integer> rightList) {
        // Count occurrences in right list
        Map<Integer, Integer> rightCounts = new HashMap<>();
        for (int num : rightList) {
            rightCounts.put(num, rightCounts.getOrDefault(num, 0) + 1);
        }

        // Calculate similarity score
        int similarityScore = 0;
        for (int num : leftList) {
            similarityScore += num * rightCounts.getOrDefault(num, 0);
        }

        return similarityScore;
    }

    public static void main(String[] args) {
        try {
            InputLists input = readInput();
            System.out.println("Part 1: " + part1(input.left, input.right));
            System.out.println("Part 2: " + part2(input.left, input.right));
        } catch (IOException e) {
            System.err.println("Error reading input: " + e.getMessage());
            System.exit(1);
        }
    }
}
